04-onion-routing.md

```diff
@@ -367,6 +379,112 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+The `encrypted_recipient_data` is a TLV stream, encrypted for a given blinded
+node, that may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_recipient_data_tlv`
+2. types:
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
+
+- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
+  - `e(0) <- {0;1}^256`
+  - `E(0) = e(0) * G`
+  - For every node `N(i)` in the route:
+    - let `P(i) = k(i) * G` be `N(i)`'s `node_id` (`k(i)` is `N(i)`'s private key)
+    - `ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))` (shared secret known only by `N(r)` and `N(i)`)
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = H(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = H(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
+- MAY create an `encrypted_recipient_data` for each blinded node `B(i)`
+- If it creates `encrypted_recipient_data`:
+  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
+  - MUST use `E(i)` in ChaCha20-Poly1305's associated authenticated data
+- MUST communicate the blinded node IDs `B(i)` and `encrypted_recipient_data(i)` to the sender
+- MUST communicate the real node ID of the introduction point `N(0)` to the sender
+- MUST communicate the first ephemeral key `E(0)` to the sender
+
+The sender:
+
+- MUST provide the `encrypted_recipient_data` field for each node in the blinded
+  route in their respective onion payloads
+- MUST communicate `E(0)` to the introduction point
+
+The introduction point:
+
+- MUST compute:
+  - `ss(0) = H(k(0) * E(0))`
+  - `rho(0) = HMAC256("rho", ss(0))`
+  - `E(1) = H(E(0) || ss(0)) * E(0)`
+- If an `encrypted_recipient_data` field is provided:
+  - MUST decrypt it using `rho(0)` and `E(0)`
+  - SHOULD use the decrypted fields to locate the next node
+- MUST forward the onion and communicate `E(1)` to the next node
+
+An intermediate node in the blinded route:
+
+- MUST receive `E(i)` from the previous node
+- MUST compute:
+  - `ss(i) = H(k(i) * E(i))`
+  - `b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)`
+  - `rho(i) = HMAC256("rho", ss(i))`
+  - `E(i+1) = H(E(i) || ss(i)) * E(i)`
+- MUST use `b(i)` instead of its private key `k(i)` to decrypt the onion
+- If an `encrypted_recipient_data` field is provided:
+  - MUST decrypt it using `rho(i)` and `E(i)`
+  - SHOULD use the decrypted fields to locate the next node
+- MUST forward the onion and communicate `E(i+1)` to the next node
+
+The recipient:
+
+- MUST receive `E(r)` from the previous node
+- MUST verify that `E(r)` matches a blinded route it created
```

Comment:
You can't really implement the first point?

And "MUST verify" is a spec anti-pattern, since it's not an action.
```
MUST ignore if the `E(r)` does not match a blinded route it created.
```

This leads to the requirement that an implementation must remember every blinded route it ever created.  Hence you really need to use a secret in `encrypted_recipient_data` IMHO.

Reply:
Good point. I'll update it to at least allow both options and clarify.

Reply:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/765/commits/cc855c754bca1ac08e8efc9f27a3c4b86773d953